src/System.Private.CoreLib/shared/System/MemoryExtensions.Trim.cs

```diff
@@ -0,0 +1,817 @@
+            return span.Slice(start, length);
+        }
+
+        /// <summary>
+        /// Removes all leading occurrences of a specified element from the span.
+        /// </summary>
+        /// <param name="span">The source span from which the element is removed.</param>
+        /// <param name="trimElement">The specified element to look for and remove.</param>
+        public static ReadOnlySpan<T> TrimStart<T>(this ReadOnlySpan<T> span, T trimElement)
+            where T : IEquatable<T>
+            => span.Slice(ClampStart(span, trimElement));
+
+        /// <summary>
+        /// Removes all trailing occurrences of a specified element from the span.
+        /// </summary>
+        /// <param name="span">The source span from which the element is removed.</param>
+        /// <param name="trimElement">The specified element to look for and remove.</param>
+        public static ReadOnlySpan<T> TrimEnd<T>(this ReadOnlySpan<T> span, T trimElement)
+            where T : IEquatable<T>
+            => span.Slice(0, ClampEnd(span, 0, trimElement));
+
+        /// <summary>
+        /// Delimits all leading occurrences of a specified element from the span.
+        /// </summary>
+        /// <param name="span">The source span from which the element is removed.</param>
+        /// <param name="trimElement">The specified element to look for and remove.</param>
+        private static int ClampStart<T>(ReadOnlySpan<T> span, T trimElement)
+            where T : IEquatable<T>
+        {
+            int start = 0;
+
+            if (trimElement != null)
+            {
+                for (; start < span.Length; start++)
+                {
+                    if (!trimElement.Equals(span[start]))
+                    {
+                        break;
+                    }
+                }
+            }
+            else
+            {
+                for (; start < span.Length; start++)
+                {
+                    if (span[start] != null)
+                    {
+                        break;
+                    }
+                }
+            }
+
+            return start;
+        }
+
+        /// <summary>
+        /// Delimits all trailing occurrences of a specified element from the span.
+        /// </summary>
+        /// <param name="span">The source span from which the element is removed.</param>
+        /// <param name="start">The start index from which to being searching.</param>
+        /// <param name="trimElement">The specified element to look for and remove.</param>
+        private static int ClampEnd<T>(ReadOnlySpan<T> span, int start, T trimElement)
+            where T : IEquatable<T>
+        {
+            // Initially, start==len==0. If ClampStart trims all, start==len
+            Debug.Assert((uint)start <= span.Length);
+
+            int end = span.Length - 1;
+
+            if (trimElement != null)
+            {
+                for (; end >= start; end--)
```

Comment:
Doesn't this >= instead of > mean we'll often compare an element twice in the Trim case? We'll move start along until it's no longer equal, which means we've checked that start isn't equal, and then here we check it again?

Reply:
Yes and no, there's 2 cases:
* `TrimEnd` calls `ClampEnd`, passing in `start == 0`, so we do need `end >= start`
* `Trim` calls `ClampStart` then `ClampEnd`, passing in `start == N`, so we need `end > start`
I think I should rather pass in `start + 1` from the relevant callsites

Reply:
Hm. I will look at this again later. I think there might be a bug still.

Reply:
No it's all good, though it's worth noting there are several interesting cases.
* `TrimEnd` calls `ClampEnd(start:0)`
* `Trim` calls `ClampStart`:
  * It `break`s immediately, returning `start==0`, then call `ClampEnd` with `end >= 1`
  * It `break`s soon, returning `start==1..Len-1`, then call `ClampEnd` with `end >= 2..Len-1`
  * It does not `break`, returning `start==Len`, then call `ClampEnd` with `end >= Len-1`

In all cases `end >= start` works fine, assuming we pass `start + 1` from the `Trim` callsites only.

Reply:
Is this something we need to apply to the string implementation?
https://github.com/dotnet/coreclr/blob/5831fe40394fcdc34fa161041956d379db433dd6/src/System.Private.CoreLib/shared/System/String.Manipulation.cs#L1888

Reply:
Do you want me to include the string fixes in this PR?

Reply:
Actually the fix is not quite as simple in those locations:
```csharp
            int start = 0;
// If we never enter this block ...
            if (trimType != TrimType.Tail)
            {
                for (start = 0; start < Length; start++)
                {...}
            }

// But we do enter this one ...
            if (trimType != TrimType.Head)
            {
// Then start == 0, and we thus need to check end >= start
                for (end = Length - 1; end >= start; end--)
                {...}
            }
```
In other words the quick fix works here iff the first block is entered

Reply:
We could however make this fix instead:
```csharp
            int start = 0;

            if (trimType != TrimType.Tail)
            {
                for (start = 0; start < Length; start++)
                {
                    if (!char.IsWhiteSpace(this[start]))
                    {
                        start++; // <-------------
                        break;
                    }
                }
            }

            if (trimType != TrimType.Head)
            {
                for (end = Length - 1; end >= start; end--)
                {...
```

Reply:
I think I would rather log this as a separate perf bug; I don't want to touch existing code in the same PR as new code. And for this particular one I want to unit it, since it looks like the final length calculation could now be wrong with the extra `++`

Reply:
> I think I would rather log this as a separate perf bug; I don't want to touch existing code in the same PR as new code.

Agreed. I wasn't suggesting making the fix here, just brought up the question since the issue looked quite similar :)

In fact, we should definitely not change the string impl in this PR.